## Final Project Submission


* Student name: Rachael Nduta Ngari
* Student pace: full time
* Scheduled project review date/time: 3/28/2025
* Instructor name: Diana Mongina
* Blog post URL:


## 1. Exploratory data Analysis (EDA) and Data cleaning

The cell below imports all necessary libraries and loads the `Aviation Data` dataset as `df`.

In [1]:
import pandas as pd  
import numpy as np  
import matplotlib.pyplot as plt  
import seaborn as sns  

%matplotlib inline  
df = pd.read_csv('Aviation_Data.csv', low_memory=False) 

**Getting the Gist of Our Data**


Before diving into analysis, let's first get the **gist** of our dataset. We'll explore its structure

In [2]:
df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90348 entries, 0 to 90347
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      90348 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88663 non-null  object 
 6   Latitude                34382 non-null  object 
 7   Longitude               34373 non-null  object 
 8   Airport.Code            50132 non-null  object 
 9   Airport.Name            52704 non-null  object 
 10  Injury.Severity         87889 non-null  object 
 11  Aircraft.damage         85695 non-null  object 
 12  Aircraft.Category       32287 non-null  object 
 13  Registration.Number     87507 non-null  object 
 14  Make                    88826 non-null

In [3]:
df.shape  

(90348, 31)

There are **31 valuables** being investigated and **90,348 accidents**, this is without removing duplicates.


In [4]:
df.columns

Index(['Event.Id', 'Investigation.Type', 'Accident.Number', 'Event.Date',
       'Location', 'Country', 'Latitude', 'Longitude', 'Airport.Code',
       'Airport.Name', 'Injury.Severity', 'Aircraft.damage',
       'Aircraft.Category', 'Registration.Number', 'Make', 'Model',
       'Amateur.Built', 'Number.of.Engines', 'Engine.Type', 'FAR.Description',
       'Schedule', 'Purpose.of.flight', 'Air.carrier', 'Total.Fatal.Injuries',
       'Total.Serious.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured',
       'Weather.Condition', 'Broad.phase.of.flight', 'Report.Status',
       'Publication.Date'],
      dtype='object')

In [5]:
df.describe()

Number.of.Engines  Total.Fatal.Injuries  Total.Serious.Injuries  \
count       82805.000000          77488.000000            76379.000000   
mean            1.146585              0.647855                0.279881   
std             0.446510              5.485960                1.544084   
min             0.000000              0.000000                0.000000   
25%             1.000000              0.000000                0.000000   
50%             1.000000              0.000000                0.000000   
75%             1.000000              0.000000                0.000000   
max             8.000000            349.000000              161.000000   

       Total.Minor.Injuries  Total.Uninjured  
count          76956.000000     82977.000000  
mean               0.357061         5.325440  
std                2.235625        27.913634  
min                0.000000         0.000000  
25%                0.000000         0.000000  
50%                0.000000         1.000000  
75%                0.000000         2.000000  
max              380.000000       699.000000

Here, we can see the **average**, **minimum**, **maximum**, and **total** number of people who **died**, were **seriously injured**, **uninjured**, and had **minor injuries**, as well as the **total number of engines**.


In [6]:
df.dtypes

Event.Id                   object
Investigation.Type         object
Accident.Number            object
Event.Date                 object
Location                   object
Country                    object
Latitude                   object
Longitude                  object
Airport.Code               object
Airport.Name               object
Injury.Severity            object
Aircraft.damage            object
Aircraft.Category          object
Registration.Number        object
Make                       object
Model                      object
Amateur.Built              object
Number.of.Engines         float64
Engine.Type                object
FAR.Description            object
Schedule                   object
Purpose.of.flight          object
Air.carrier                object
Total.Fatal.Injuries      float64
Total.Serious.Injuries    float64
Total.Minor.Injuries      float64
Total.Uninjured           float64
Weather.Condition          object
Broad.phase.of.flight      object
Report.Status 

**Data Cleaning**


Structuring, Detecting, and correcting errors, inconsistencies, and inaccuracies in a dataset to improve its quality and reliability.  


Remove Duplicates and Drop Duplicate `Accident.Number` and `Event.Id`

In [7]:
df.duplicated().sum()

1390

In [8]:
df.drop_duplicates(inplace=True)

In [9]:
df.shape

(88958, 31)

In [10]:
df[df.duplicated(subset=['Accident.Number'], keep=False)]

Event.Id Investigation.Type Accident.Number  Event.Date  \
64030             NaN         25-09-2020             NaN         NaN   
69843             NaN         14-06-2021             NaN         NaN   
74420             NaN         03-11-2020             NaN         NaN   
74946             NaN         24-02-2021             NaN         NaN   
75894             NaN         02-02-2021             NaN         NaN   
...               ...                ...             ...         ...   
90255  20221121106336           Accident      WPR23LA041  2022-11-18   
90256  20221122106340           Incident      DCA23WA071  2022-11-18   
90257  20221122106340           Incident      DCA23WA071  2022-11-18   
90272  20221123106354           Accident      WPR23LA045  2022-11-22   
90273  20221123106354           Accident      WPR23LA045  2022-11-22   

            Location        Country Latitude Longitude Airport.Code  \
64030            NaN            NaN      NaN       NaN          NaN   
69843            NaN            NaN      NaN       NaN          NaN   
74420            NaN            NaN      NaN       NaN          NaN   
74946            NaN            NaN      NaN       NaN          NaN   
75894            NaN            NaN      NaN       NaN          NaN   
...              ...            ...      ...       ...          ...   
90255  Las Vegas, NV  United States  361239N  1151140W          VGT   
90256    Marrakech,         Morocco      NaN       NaN          NaN   
90257    Marrakech,         Morocco      NaN       NaN          NaN   
90272  San Diego, CA  United States  323414N  1165825W          SDM   
90273  San Diego, CA  United States  323414N  1165825W          SDM   

                        Airport.Name  ...          Purpose.of.flight  \
64030                            NaN  ...                        NaN   
69843                            NaN  ...                        NaN   
74420                            NaN  ...                        NaN   
74946                            NaN  ...                        NaN   
75894                            NaN  ...                        NaN   
...                              ...  ...                        ...   
90255                NORTH LAS VEGAS  ...              Instructional   
90256                            NaN  ...                        NaN   
90257                            NaN  ...                        NaN   
90272  Brown Field Municipal Airport  ...              Instructional   
90273  Brown Field Municipal Airport  ...  Public Aircraft - Federal   

               Air.carrier Total.Fatal.Injuries Total.Serious.Injuries  \
64030                  NaN                  NaN                    NaN   
69843                  NaN                  NaN                    NaN   
74420                  NaN                  NaN                    NaN   
74946                  NaN                  NaN                    NaN   
75894                  NaN                  NaN                    NaN   
...                    ...                  ...                    ...   
90255   702 HELICOPTER INC                  0.0                    0.0   
90256      British Airways                  0.0                    0.0   
90257  Valair Private Jets                  0.0                    0.0   
90272      HeliStream Inc.                  0.0                    0.0   
90273            U.S. Navy                  0.0                    0.0   

      Total.Minor.Injuries Total.Uninjured Weather.Condition  \
64030                  NaN             NaN               NaN   
69843                  NaN             NaN               NaN   
74420                  NaN             NaN               NaN   
74946                  NaN             NaN               NaN   
75894                  NaN             NaN               NaN   
...                    ...             ...               ...   
90255                  0.0             3.0               VMC   
90256                  0.0   

In [11]:
df = df.drop_duplicates(subset=['Accident.Number'])

In [12]:
df[df.duplicated(subset=['Event.Id], keep=False)]

Event.Id Investigation.Type Accident.Number  Event.Date  \
117    20020917X01908           Accident     DCA82AA012B  1982-01-19   
118    20020917X01908           Accident     DCA82AA012A  1982-01-19   
153    20020917X02259           Accident     LAX82FA049A  1982-01-23   
158    20020917X02400           Accident     MIA82FA038B  1982-01-23   
159    20020917X02400           Accident     MIA82FA038A  1982-01-23   
...               ...                ...             ...         ...   
63823  20071227X02001           Accident     NYC08FA054A  2007-12-08   
63841  20080213X00187           Accident     NYC08LA062B  2007-12-14   
63842  20080213X00187           Accident     NYC08LA062A  2007-12-14   
63897  20080128X00107           Incident     OPS08IA004B  2007-12-27   
63898  20080128X00107           Incident     OPS08IA004A  2007-12-27   

                 Location        Country   Latitude   Longitude Airport.Code  \
117          ROCKPORT, TX  United States        NaN         NaN          RKP   
118          ROCKPORT, TX  United States        NaN         NaN          RKP   
153       VICTORVILLE, CA  United States        NaN         NaN          NaN   
158    NEWPORT RICHEY, FL  United States        NaN         NaN          NaN   
159    NEWPORT RICHEY, FL  United States        NaN         NaN          NaN   
...                   ...            ...        ...         ...          ...   
63823        Parkland, FL  United States  26.347222  -80.315278          NaN   
63841        NEW YORK, NY  United States  40.658333  -73.786666          JFK   
63842        NEW YORK, NY  United States  40.658333  -73.786666          JFK   
63897       Teterboro, NJ  United States        NaN         NaN          NaN   
63898       Teterboro, NJ  United States        NaN         NaN          NaN   

                        Airport.Name  ...    Purpose.of.flight  \
117           ARANSAS COUNTY AIRPORT  ...             Personal   
118           ARANSAS COUNTY AIRPORT  ...  Executive/corporate   
153                              NaN  ...             Personal   
158                              NaN  ...             Personal   
159                              NaN  ...             Personal   
...                              ...  ...                  ...   
63823                            NaN  ...             Personal   
63841  John F. Kennedy International  ...          Positioning   
63842  John F. Kennedy International  ...          Positioning   
63897                            NaN  ...                  NaN   
63898                            NaN  ...              Unknown   

              Air.carrier Total.Fatal.Injuries Total.Serious.Injuries  \
117                   NaN                  3.0                    0.0   
118                   NaN                  3.0                    0.0   
153                   NaN                  2.0                    0.0   
158                   NaN                  0.0                    0.0   
159                   NaN                  0.0                    0.0   
...                   ...                  ...                    ...   
63823                 NaN                  2.0                    NaN   
63841                 NaN                  NaN                    NaN   
63842  Air China Airlines                  NaN                    NaN   
63897                 NaN                  NaN                    NaN   
63898                 NaN                  NaN                    NaN   

      Total.Minor.Injuries Total.Uninjured Weather.Condition  \
117                    0.0             0.0               IMC   
118                    0.0             0.0               IMC   
153                    4.0             0.0               VMC   
158                    0.0             3.0               VMC   
159                    0.0             3.0               VMC   
...                    ...             ...               ...   
63823                  NaN             NaN               VMC   
63841

In [13]:
df = df.drop_duplicates(subset=['Event.Id'])

In [14]:
df.shape

(87952, 31)

Set `Event.ID` as the Index

In [15]:
df['Accident.Number'].isna().sum()

1

In [17]:
df = df.dropna(subset=["Accident.Number"])

In [18]:
df['Event.Id'].isna().sum()

0

In [19]:
df=df.set_index('Event.Id')

In [21]:
df.shape

(87951, 30)

In [22]:
df.isna().sum()

Investigation.Type            0
Accident.Number               0
Event.Date                    0
Location                     52
Country                     222
Latitude                  53739
Longitude                 53748
Airport.Code              38467
Airport.Name              35920
Injury.Severity             990
Aircraft.damage            3103
Aircraft.Category         55770
Registration.Number        1350
Make                         63
Model                        92
Amateur.Built               100
Number.of.Engines          6027
Engine.Type                7043
FAR.Description           56036
Schedule                  75591
Purpose.of.flight          6122
Air.carrier               71418
Total.Fatal.Injuries      11267
Total.Serious.Injuries    12322
Total.Minor.Injuries      11760
Total.Uninjured            5863
Weather.Condition          4473
Broad.phase.of.flight     27114
Report.Status              6364
Publication.Date          15057
dtype: int64

Checking the Percentage of Missing Data in Columns with the Highest Missing Values and Dropping Those Above 75%  


In [23]:
df['Schedule'].isna().mean() * 100

85.94672033291263

In [24]:
df['Air.carrier'].isna().mean() * 100

81.20203295016543

In [25]:
df = df.drop(columns=['Schedule'])

df = df.drop(columns=['Air.carrier'])

In [26]:
df.shape

(87951, 28)

In [27]:
df['Investigation.Type'].value_counts()

Investigation.Type
Accident    84190
Incident     3761
Name: count, dtype: int64

In [28]:
df_incidents=df[df["Investigation.Type"] == "Incident"]

In [29]:
df_incidents["Injury.Severity"].value_counts()

Injury.Severity
Incident     2113
Non-Fatal     977
Minor           3
Fatal           1
Serious         1
Name: count, dtype: int64

Remove rows with incidences, we are not going to use them and most of the causes do not have injuriies or fatalities 

In [30]:
df = df[df["Investigation.Type"] != "Incident"]

In [31]:
df.shape


(84190, 28)

In [32]:
df.columns

Index(['Investigation.Type', 'Accident.Number', 'Event.Date', 'Location',
       'Country', 'Latitude', 'Longitude', 'Airport.Code', 'Airport.Name',
       'Injury.Severity', 'Aircraft.damage', 'Aircraft.Category',
       'Registration.Number', 'Make', 'Model', 'Amateur.Built',
       'Number.of.Engines', 'Engine.Type', 'FAR.Description',
       'Purpose.of.flight', 'Total.Fatal.Injuries', 'Total.Serious.Injuries',
       'Total.Minor.Injuries', 'Total.Uninjured', 'Weather.Condition',
       'Broad.phase.of.flight', 'Report.Status', 'Publication.Date'],
      dtype='object')

In [33]:
df.head()

Investigation.Type Accident.Number  Event.Date  \
Event.Id                                                        
20001218X45444           Accident      SEA87LA080  1948-10-24   
20001218X45447           Accident      LAX94LA336  1962-07-19   
20061025X01555           Accident      NYC07LA005  1974-08-30   
20001218X45448           Accident      LAX96LA321  1977-06-19   
20041105X01764           Accident      CHI79FA064  1979-08-02   

                       Location        Country   Latitude   Longitude  \
Event.Id                                                                
20001218X45444  MOOSE CREEK, ID  United States        NaN         NaN   
20001218X45447   BRIDGEPORT, CA  United States        NaN         NaN   
20061025X01555    Saltville, VA  United States  36.922223  -81.878056   
20001218X45448       EUREKA, CA  United States        NaN         NaN   
20041105X01764       Canton, OH  United States        NaN         NaN   

               Airport.Code Airport.Name Injury.Severity  ... FAR.Description  \
Event.Id                                                  ...                   
20001218X45444          NaN          NaN        Fatal(2)  ...             NaN   
20001218X45447          NaN          NaN        Fatal(4)  ...             NaN   
20061025X01555          NaN          NaN        Fatal(3)  ...             NaN   
20001218X45448          NaN          NaN        Fatal(2)  ...             NaN   
20041105X01764          NaN          NaN        Fatal(1)  ...             NaN   

               Purpose.of.flight Total.Fatal.Injuries Total.Serious.Injuries  \
Event.Id                                                                       
20001218X45444          Personal                  2.0                    0.0   
20001218X45447          Personal                  4.0                    0.0   
20061025X01555          Personal                  3.0                    NaN   
20001218X45448          Personal                  2.0                    0.0   
20041105X01764          Personal                  1.0                    2.0   

               Total.Minor.Injuries Total.Uninjured  Weather.Condition  \
Event.Id                                                                 
20001218X45444                  0.0             0.0                UNK   
20001218X45447                  0.0             0.0                UNK   
20061025X01555                  NaN             NaN                IMC   
20001218X45448                  0.0             0.0                IMC   
20041105X01764                  NaN             0.0                VMC   

               Broad.phase.of.flight   Report.Status Publication.Date  
Event.Id                                                               
20001218X45444                Cruise  Probable Cause              NaN  
20001218X45447               Unknown  Probable Cause       19-09-1996  
20061025X01555                Cruise  Probable Cause       26-02-2007  
20001218X45448                Cruise  Probable Cause       12-09-2000  
20041105X01764              Approach  Probable Cause       16-04-1980  

[5 rows x 28 columns]

Convert **"Event.Date"** to Datetime Format

In [38]:
df["Event.Date"] = pd.to_datetime(df["Event.Date"])

In [39]:
df["Event.Date"].isna().sum()


0

Cleaning location 

In [52]:
df["Location"]=df["Location"].str.strip()

In [53]:
df["Location"].isna().sum()


40

In [56]:
df["Country"].isna().sum()

204

In [57]:
df["Country"].value_counts()

Country
United States     79112
Brazil              330
Canada              308
Mexico              289
United Kingdom      219
                  ...  
Rwanda                1
Palau                 1
Libya                 1
Guinea                1
Great Britain         1
Name: count, Length: 209, dtype: int64

In [58]:
df["Latitude"].isna().sum()

50944

In [60]:
df["Longitude"].isna().sum()

50953

In [62]:
df['Aircraft.damage'].value_counts()

Aircraft.damage
Substantial    63599
Destroyed      18394
Minor            655
Unknown           97
Name: count, dtype: int64

In [65]:
df['Aircraft.damage']=df['Aircraft.damage'].str.strip()

In [66]:
df['Aircraft.damage'].isna().sum()

1445